chalk-integration/src/program.rs

```diff
@@ -422,14 +422,36 @@ impl RustIrDatabase<ChalkIr> for Program {
             .collect()
     }
 
-    fn impl_provided_for(&self, auto_trait_id: TraitId<ChalkIr>, adt_id: AdtId<ChalkIr>) -> bool {
+    fn impl_provided_for(
+        &self,
+        auto_trait_id: TraitId<ChalkIr>,
+        app_ty: &ApplicationTy<ChalkIr>,
+    ) -> bool {
         let interner = self.interner();
-        // Look for an impl like `impl Send for Foo` where `Foo` is
-        // the ADT.  See `push_auto_trait_impls` for more.
-        self.impl_data.values().any(|impl_datum| {
-            impl_datum.trait_id() == auto_trait_id
-                && impl_datum.self_type_adt_id(interner) == Some(adt_id)
-        })
+
+        // an iterator containing the `ApplicationTy`s which have an impl for the trait `auto_trait_id`.
+        let mut impl_app_tys = self.impl_data.values().filter_map(|impl_datum| {
+            if impl_datum.trait_id() != auto_trait_id {
+                return None;
+            }
+
+            let ty = impl_datum
+                .binders
+                .skip_binders()
+                .trait_ref
+                .self_type_parameter(interner);
+            match ty.data(interner) {
+                TyData::Apply(app) => Some(app.clone()),
+                _ => None,
+            }
+        });
+
+        // we only compare the `TypeName`s as
```

Comment:
I wonder if it's worth passing the substitution to this function, even if unused? Theoretically, if we wanted to match rustc's behavior in the integration, we could.

Reply:
I think the answer is no. Interesting thought, but let's not overcomplicate.